lib/ansible/modules/cloud/amazon/route53_zone.py

```diff
@@ -216,54 +244,44 @@ def create_public(conn, matching_zones, zone_in, record):
-        if isinstance(zone_details['VPCs'], dict):
-            if zone_details['VPCs']['VPC']['VPCId'] == vpc_id and vpc_region == zone_details['VPCs']['VPC']['VPCRegion']:
-                conn.delete_hosted_zone(z)
-                changed = True
-                msg = "Successfully deleted %s" % matching_zones[z]
-                break
-            else:
-                changed = False
+        result = client.get_hosted_zone(Id=z['Id'])
+        zone_details = result['HostedZone']
+        vpc_details = result['VPCs']
+        if isinstance(vpc_details, dict):
+            if vpc_details['VPC']['VPCId'] == vpc_id and vpc_region == vpc_details['VPC']['VPCRegion']:
+                client.delete_hosted_zone(Id=z['Id'])
+                return True, "Successfully deleted %s" % zone_details['Name']
         else:
-            if vpc_id in [v['VPCId'] for v in zone_details['VPCs']] and vpc_region in [v['VPCRegion'] for v in zone_details['VPCs']]:
-                conn.delete_hosted_zone(z)
-                changed = True
-                msg = "Successfully deleted %s" % matching_zones[z]
-                break
-            else:
-                changed = False
-    if not changed:
-        msg = "The vpc_id and the vpc_region do not match a private hosted zone."
+            if vpc_id in [v['VPCId'] for v in vpc_details] and vpc_region in [v['VPCRegion'] for v in vpc_details]:
+                client.delete_hosted_zone(Id=z['Id'])
+                return True, "Successfully deleted %s" % zone_details['Name']
 
-    return changed, msg
+    return False, "The vpc_id and the vpc_region do not match a private hosted zone."
 
 
-def delete_public(conn, matching_zones):
+def delete_public(client, matching_zones):
     if len(matching_zones) > 1:
         changed = False
         msg = "There are multiple zones that match. Use hosted_zone_id to specify the correct zone."
     else:
-        for z in matching_zones:
-            conn.delete_hosted_zone(z)
-            changed = True
-            msg = "Successfully deleted %s" % matching_zones[z]
+        client.delete_hosted_zone(Id=matching_zones[0]['Id'])
+        changed = True
+        msg = "Successfully deleted %s" % matching_zones[0]['Id']
     return changed, msg
 
 
-def delete_hosted_id(conn, hosted_zone_id, matching_zones):
+def delete_hosted_id(client, hosted_zone_id, matching_zones):
     if hosted_zone_id == "all":
         deleted = []
         for z in matching_zones:
-            deleted.append(z)
-            conn.delete_hosted_zone(z)
+            deleted.append(z['Id'])
+            client.delete_hosted_zone(Id=z['Id'])
         changed = True
         msg = "Successfully deleted zones: %s" % deleted
-    elif hosted_zone_id in matching_zones:
-        conn.delete_hosted_zone(hosted_zone_id)
+    elif hosted_zone_id in [zo['Id'].replace('/hostedzone/', '') for zo in matching_zones]:
+        client.delete_hosted_zone(Id=hosted_zone_id)
```

Comment:
All client calls should have exception handling.

https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/GUIDELINES.md#using-fail_json_aws